### PR TITLE
Use newer mm-data encoder interface when possible

### PR DESCRIPTION
**Pull request by wdhongtw**
# Description

If user explicitly specifies using "Encoder CUDA Graph" feature (and the model does support the interface),
use the new interface to achieve better performance and avoid runtime re-compilation that slows down online-inference services.

- Use `SupportsEncoderCudaGraph.encoder_cudagraph_forward` instead of `SupportsMultiModal.embed_multimodal`
- Implement pre-compilation for supported token budgets
- Split mm-data into multiple batches when necessary
- Move the encoder CUDA graph logic into `MultiModalManager` so it can also be leveraged by the JAX path in the future
- Extract mm encoder manager into a separate file (`tpu_inference/models/vllm/mm_encoder_manager.py`) with unit tests
- Patch encoder forward and corresponding pre-compile function in `TPURunner` when appropriate

Corp internal doc: <a href="https://docs.google.com/document/d/10cYjQllhOhRPhtzuKLxS4G6HbpHz9lu56P6KA4zZve8/edit?usp=sharing">link</a>

### Performance

| Metric | Before | After |
|--------|--------|-------|
| Throughput (req/s) | 0.78 | 3.05 |

# Tests

A command to trigger corresponding logic.

```shell
MODEL_IMPL_TYPE=vllm python examples/generate/multimodal/vision_language_offline.py --model-type=qwen3_vl --num-prompts=1 --modality=image --time-generate  -tp=4 --enable-vit-cuda-graph
```

Also, for benchmarking.

```shell
MODEL_IMPL_TYPE=vllm vllm serve Qwen/Qwen3-VL-2B-Instruct \
  --max-model-len 8192 --disable-uvicorn-access-log \
  --max-num-batched-tokens 4096 --max-num-seqs 256 \
  --tensor-parallel-size 8 --trust-remote-code \
  --limit-mm-per-prompt '{"image": 8}' \
  --no-enable-prefix-caching --mm-processor-cache-gb=0 --gpu-memory-utilization 0.9 \
  --disable-chunked-mm-input \
  --mm-processor-kwargs '{"max_pixels": 512000}' \
  --compilation-config '{"cudagraph_mm_encoder": true}'

client:
time python scripts/vllm/benchmarking/benchmark_serving.py \
  --backend vllm-chat \
  --endpoint /v1/chat/completions \
  --model Qwen/Qwen3-VL-2B-Instruct \
  --dataset-name mmmu_pro \
  --mmmu-pro-subset 'standard (10 options)' \
  --mmmu-pro-output-len=1 \
  --run-eval \
  --num-prompts 1730 \
  --max-concurrency 32
```

Unit tests for the extracted mm encoder manager:

```shell
pytest tests/models/vllm/test_mm_encoder_manager.py
```

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have made or will make corresponding changes to any relevant documentation.
- [ ] I have reviewed the uLLM modeling code checklist: https://docs.google.com/document/d/1DGQBVvr2bh4G8tBUO1YH8pO7Dd_myw5rfEMfVDymEk8/edit?resourcekey=0-V7MGHu3aQjJH6YrI3-y8Hg&tab=t.t91cyovog2mr#heading=h.cqdzv8mlszca
- [ ] I have received at least 1 readability approval and 1 correctness approval.

### CODE REVIEW

File: tpu_inference/models/vllm/vllm_model_wrapper.py [hunk 1]
```diff
@@ -69,137 +75,154 @@
...
+                    dtype=self.dtype,
+                ).values
+
+            return inputs
+
+        self.by_budget = {b: build_inputs(b) for b in self.token_budgets}
+        """The "input tensors" for each token budget."""
+
+    @staticmethod
+    def _generate_budgets(min_budget: int, max_budget: int) -> list[int]:
```
Comment (lk-chen) [inline tpu_inference/models/vllm/vllm_model_wrapper.py]: this is identical to [vllm](https://github.com/vllm-project/vllm/blob/ba94a3b9989666f950e1f784d18f2033c63c6cad/vllm/v1/worker/encoder_cudagraph.py#L171-L181) any reason not re-use that?

Comment (wdhongtw) [reply tpu_inference/models/vllm/vllm_model_wrapper.py]: Yes, it's identical. Here I try to only using public interface, if possible.

File: tpu_inference/models/vllm/vllm_model_wrapper.py [hunk 2]
```diff
@@ -69,137 +75,154 @@
...
-    def compute_logits(self, hidden_state: torch.Tensor) -> torch.Tensor:
-        return self.vllm_model.compute_logits(hidden_state)
+        self.config = model.get_encoder_cudagraph_config()
+        self.dtype = vllm_config.model_config.dtype
+
+        comp_config = vllm_config.compilation_config
+        multimodal_config = vllm_config.model_config.multimodal_config
+        assert multimodal_config is not None
+
+        def build_budgets_and_batch() -> tuple[list[int], int]:
```
Comment (lk-chen) [inline tpu_inference/models/vllm/vllm_model_wrapper.py]: these could be class-level function instead of nested function

File: tpu_inference/models/vllm/vllm_model_wrapper.py [hunk 3]
```diff
@@ -69,137 +75,154 @@
...
+
+    # None does occur as some field values of mm_kwargs in some model,
+    # like Qwen 3 VL, during prepare_encoder_cudagraph_replay_buffers.
+    if v is None:
+        return None
+
+    return torch_view(t2j(v, use_dlpack=False))
+
+
+class EncoderManager:
```
Comment (lk-chen) [inline tpu_inference/models/vllm/vllm_model_wrapper.py]: let's move it to a separate file and add unit test

**Review by lk-chen — COMMENTED** [review]
Maybe tidy up the PR by decoupling from #2762 ?

File: tpu_inference/models/vllm/vllm_model_wrapper.py (hunk 3 above)
Comment (wdhongtw) [reply tpu_inference/models/vllm/vllm_model_wrapper.py]: No problem, will do.

**Review by lk-chen — COMMENTED** [review]
LGTM on [mm_encoder_manager.py‎](https://github.com/vllm-project/tpu-inference/pull/2804/changes#diff-957424c70e02baae6ef985866903f5b1ebdf1e4b6eb6aaf5254808410b33364d), let's wait for #2762

**Comment by wdhongtw** [comment]
Update out of channel suggestion from @lk-chen at 5/10

> Move the logic from vllm_model_wrapper.py into multimodal_manager.py, so it also be used by jax path later.

**Comment by wdhongtw** [comment]
Move the decision of using new mm encoder interface into MultiModalManager.
Try to patch the encoder forward and corresponding pre-compile function in TPURunner if appropriate.

@lk-chen

---

The original impl that only changes VllmModelWrapper (torchax path) is also all green in CI.
For tracking purpose, it's now recorded at personal fork https://github.com/wdhongtw/tpu-inference/releases/tag/wdhongtw%2Fmm-rework-0610.

**Comment by wdhongtw** [comment]
CI jobs failed due to disk space full.

> failed to register layer: write /usr/local/lib/python3.12/site-packages/triton/backends/nvidia/bin/ptxas-blackwell: no space left on device

Force push to retry again.

File: tpu_inference/runner/multimodal_manager.py [hunk 4]
```diff
@@ -270,3 +307,248 @@ def gather_mm_embeddings(
...
+        embed_multimodal_fn,
+        precompile_vision_encoder_fn,
+    ):
+        original_functions = (
+            embed_multimodal_fn,
+            precompile_vision_encoder_fn,
+        )
+
+        # Only support using SupportsEncoderCudaGraph in torchax path.
+        if tpu_inference.envs.MODEL_IMPL_TYPE != "vllm":
```
Comment (lk-chen) [inline tpu_inference/runner/multimodal_manager.py:322]: could be "auto", will help fix in upcoming PR

File: tpu_inference/models/vllm/mm_encoder_manager.py [hunk 5]
```diff
@@ -0,0 +1,166 @@
+# Copyright 2025 Google LLC
```
Comment (lk-chen) [inline tpu_inference/models/vllm/mm_encoder_manager.py:1]: 2026

File: tpu_inference/runner/multimodal_manager.py [hunk 6]
```diff
@@ -270,3 +307,248 @@ def gather_mm_embeddings(
...
+                    budget,
+                    budget=budget,
+                )
+
+        def embed_multimodal_graph_forward_all(
+            jax_params_and_buffers: dict[str, jax.Array],
+            **mm_kwargs: dict[str, NestedTensors],
+        ) -> list[jax.Array]:
+
+            def get_fit_val(values: list[int], value: int) -> int | None:
```
Comment (lk-chen) [inline tpu_inference/runner/multimodal_manager.py:410]: a bit concern on code quality - 3-layer nested method

**Review by lk-chen — APPROVED** [review]
better have before/after metrics shown in PR description. 

Will wait for CI finish and merge, to unblock future work

**Comment by lk-chen** [comment]
new throughput [3.05 req/s](https://buildkite.com/tpu-commons/tpu-inference-benchmark/builds/1404#019ebcf4-f2b4-466e-8064-0bcf176a35ab/L2176)
before: 0.78 req/s

**Comment by lk-chen** [comment]
@copilot show per-file owner (don't at them, just list user name)

**Review by jrplatin — APPROVED** [review]
